File: Maximum Frequency of an Element After Performing Operations I/Maximum Frequency of an Element After Performing Operations I.java
// Solution for Maximum Frequency of an Element After Performing Operations I in JAVA

class Solution {
    public int maxFrequency(int[] nums, int k, int numOperations) {
        int maxVal = Arrays.stream(nums).max().getAsInt() + k;
        int[] prefix = new int[maxVal + 1];

        for (int val : nums) prefix[val]++;
        for (int i = 1; i <= maxVal; i++) prefix[i] += prefix[i - 1];

        int ans = 0;
        for (int target = 0; target <= maxVal; target++) {
            if (prefix[target] == 0) continue;

            int left = Math.max(target - k, 0);
            int right = Math.min(target + k, maxVal);

            int total = prefix[right] - (left > 0 ? prefix[left - 1] : 0);
            int currFreq = prefix[target] - (target > 0 ? prefix[target - 1] : 0);
            int convert = Math.min(total - currFreq, numOperations);

            ans = Math.max(ans, currFreq + convert);
        }
        return ans;
    }
}

File: Maximum Frequency of an Element After Performing Operations I/Maximum Frequency of an Element After Performing Operations I.py
# Solution for Maximum Frequency of an Element After Performing Operations I in PY

class Solution:
    def maxFrequency(self, nums, k, numOperations):
        max_val = max(nums) + k
        prefix = [0] * (max_val + 1)
        for v in nums:
            prefix[v] += 1
        for i in range(1, max_val + 1):
            prefix[i] += prefix[i - 1]

        ans = 0
        for target in range(max_val + 1):
            if prefix[target] == 0:
                continue
            left, right = max(target - k, 0), min(target + k, max_val)
            total = prefix[right] - (prefix[left - 1] if left > 0 else 0)
            curr = prefix[target] - (prefix[target - 1] if target > 0 else 0)
            ans = max(ans, curr + min(total - curr, numOperations))
        return ans
